osf/metrics.py

```diff
@@ -124,6 +124,10 @@ class BasePreprintMetric(MetricMixin, metrics.Metric):
     preprint_id = metrics.Keyword(index=True, doc_values=True, required=True)
     version = metrics.Keyword(index=True, doc_values=True)
     path = metrics.Text(index=True)
+    referrer = metrics.Text(index=True)
```

Comment:
from https://www.elastic.co/guide/en/elasticsearch/reference/current/text.html:
> If you need to index structured content such as email addresses, hostnames, status codes, or tags, it is likely that you should rather use a keyword field.

I think this extends to everything we're tracking here, including `path` which was initially just text, I'm going to flip all these attributes over to keywords, I think Steve just made a mistake with `path`.

Reply:
My guess is that we don't want all of them to be keyword fields, I just don't yet know enough to figure out how to query it. For example, `user_agent` is so lengthy `user_agent='Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:64.0) Gecko/20100101 Firefox/64.0'`, we would probably want partial matches.

Reply:
Regardless of how we query, I'd say all the new fields are closer to `Keywords` as described via the docs then `text` which seems to be just human-readable sentences and phrases. `count` should obviously be an integer though.

Reply:
From [keyword fields](https://www.elastic.co/guide/en/elasticsearch/reference/current/keyword.html#keyword)
> Keyword fields are only searchable by their exact value.

So I'm thinking some of these new fields should be Keyword fields but some of them should be text fields.  My main question is if it's a textfield, how do we query against it?

@brianjgeiger - can you weigh in on Keyword versus text fields?  Seems like we need to make sure we can get the info we want out of these fields.

Reply:
If we can't search on a text field, then my guess is that it is not of use to us at all, right? Unless you can aggregate based on the text field but keyword is the only thing you can search on? That _could_ be.

Reply:
I experimented some yesterday, you _can_ search on the text field, but i believe it's returning partial matches.  And since we're not returning `source`, it's hard to tell what's in the response. (I enabled source locally and verified it's doing some sort of partial match with cls.search()).filter('match', referrer="http://localhost:5000/sdfas/")

I want to make sure if we move over to keyword field that this is still useful.  For example, is a user agent so specific, that if we search for it, we'll only get one result returned?  In this case, partial matches would be better. 

Do you have an opinion @mfraezz?

Reply:
Gotcha. No, user agent would be a terrible idea for keyword, in that case. Referrer probably the same.

Reply:
version is the lastest version, but I'm thinking it might be ok to leave as keyword field, that can be an exact match. also don't know the implications of changing field type.

@Johnetordoff, probably `path`, `referrer`, `user_agent`, `origin`, and `uri` should go back to text fields, sorry my comments led you to change these! but we also as a group need to better learn how to query text fields and what we can expect in the response.

Reply:
The only reason I was thinking a number field is because then you could do aggregations based on ranges of versions, so one bucket would have 5 or fewer versions, the next would have 6-10, etc. All that being said, yeah, probably not worth it and possibly not feasible.

Reply:
that's a good point @brianjgeiger, that field is probably not useful for aggregations

Reply:
So I asked @mfraezz's opinion and he suggested we need to figure out what kinds of questions we're going to ask with this data, which is a good point. Maybe we're putting the cart before the horse.      

His example, "How many user agents of each type during [date range]?"  So our new fields, referrer, user_agent, origin, uri, what kinds of questions do we want to ask? And if we're doing aggregations, they might need to be keyword fields.   He also mentioned there might be complications with source being disabled.

Some links he sent: 
https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html
https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-significantterms-aggregation.html and https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-valuecount-aggregation.html

Reply:
So I think probably the next step @Johnetordoff, is to think about what kinds of questions we want to ask from these fields - looks like they suspect counts are coming from google scholar, so that's a starting point. 

Once we know the kind of query(s) we want to write, then we can better decide how the field should be structured.  (And I'm going to avoid weighing in anymore on this thread about changing this or that, because I'm the worst!) . Let's just back up!